src/lib/operators/insert.cpp

```diff
@@ -100,115 +124,163 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
 
   _target_table = StorageManager::get().get_table(_target_table_name);
 
+  // partitioning
+  std::shared_ptr<const AbstractPartitionSchema> target_partition_schema = _target_table->get_partition_schema();
+  std::map<RowID, PartitionID> target_partition_mapping =
+      target_partition_schema->get_mapping_to_partitions(_input_table_left());
```

Comment:
I get that you want to calculate the counts so that you can reserve the right amount of space. But what is your reason to store the mapping instead of calculating it on the fly? When you insert the data, you have to iterate over it a second time anyway - shouldn't recalculating the mapping be quite cheap? At least, I would expect it to be cheaper than maintaining this map/tree and later vectors of rows to insert.

I'll skip the remainder of this file for now.